libcontainer/nsenter/nsexec.c

```diff
@@ -1225,8 +1226,24 @@ void nsexec(void)
 			 * was broken, so we'll just do it the long way anyway.
 			 */
 			write_log(DEBUG, "unshare remaining namespace (except cgroupns)");
-			if (unshare(config.cloneflags & ~CLONE_NEWCGROUP) < 0)
-				bail("failed to unshare remaining namespaces (except cgroupns)");
+			/*
+			 * In centos 7.4/7.5/7.6, when other processes read the /proc/pid/status
+			 * file of the runc: [1: CHILD] process, and if the runc: [1: CHILD]
+			 * process happens to be in the unshare stage, the unshare syscall will
+			 * report an error EINVAL. This is because when other processes read
+			 * the status file, they will call the kernel function get_ task_ mm(),
+			 * which sets the decision condition task->mm->mm_users +1. In the
+			 * unshare syscall, mm does not meet the condition of mm<=1, and the
+			 * kernel throws an EINVAL error. For this, I think we can try again.
+			 */
```

Comment:
Doesn't this affect the rest of the `unshare()` calls here too?

Reply:
The retry mechanism only occurs when unshare fails, and the unshare successfully executed only once.

Reply:
Exactly, but my question is why don't we need the retry mechanism on the other `unshare()` calls we do here. For example, just a few lines above this, still in the runc CHILD, we unshare the userns: https://github.com/opencontainers/runc/pull/3705/files#diff-6383238247e090d88ade6343c0ef59dd09b3c10634bf0584e78445b843c55ab0R1175

There are other unshare calls too. Don't we want to cover with this retry mechanism all of them?

Reply:
@zzyyzte btw, IIRC there are more calls to unshare than these two

Reply:
I would also write a wrapper for unshare that will retry and use it from all places that call `unshare` now. That should not create any measurable overhead I guess.

Reply:
Thanks for suggestions. I create a new function to try unshare syscall.